command/healthcheck/pki.go

```diff
@@ -227,3 +227,49 @@ func pkiFetchLeaf(e *Executor, serial string, versionError func()) (bool, *PathF
 
 	return false, leafRet, leafRet.ParsedCache["certificate"].(*x509.Certificate), nil
 }
+
+func pkiFetchRoles(e *Executor, versionError func()) (bool, *PathFetch, []string, error) {
+	rolesRet, err := e.FetchIfNotFetched(logical.ListOperation, "/{{mount}}/roles")
+	if err != nil {
+		return true, nil, nil, err
+	}
+
+	if !rolesRet.IsSecretOK() {
+		if rolesRet.IsUnsupportedPathError() {
+			versionError()
+		}
+
+		return true, nil, nil, nil
+	}
+
+	if len(rolesRet.ParsedCache) == 0 {
+		var roles []string
+		for _, rawSerial := range rolesRet.Secret.Data["keys"].([]interface{}) {
+			roles = append(roles, rawSerial.(string))
+		}
+		rolesRet.ParsedCache["roles"] = roles
+	}
+
+	return false, rolesRet, rolesRet.ParsedCache["roles"].([]string), nil
+}
+
+func pkiFetchRole(e *Executor, name string, versionError func()) (bool, *PathFetch, map[string]interface{}, error) {
+	roleRet, err := e.FetchIfNotFetched(logical.ReadOperation, "/{{mount}}/roles/"+name)
+	if err != nil {
+		return true, nil, nil, err
+	}
+
+	if !roleRet.IsSecretOK() {
+		if roleRet.IsUnsupportedPathError() {
+			versionError()
+		}
+		return true, nil, nil, nil
+	}
+
+	var data map[string]interface{} = nil
```

Comment:
Any particular reason we don't cache the fetched roles?

Reply:
I don't think we'd really need to do much to cache it?

The philosophy behind the cache variable is to store parsed forms.

The role JSON response (in `secret.Data`) is about exactly what we'd store here I'd think, +/- perhaps collapsing a few `[]interface{}` directly to `[]string` or similar. 

If you have ideas of data to fetch/cache here, be curious to hear it.

Reply:
Ah thanks for the clarification, I had missed the fact that we always internally cache the response automatically.